lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java

```diff
@@ -130,10 +138,46 @@ private NettyMessagingTransport(
 
     final String host = UNKNOWN_HOST_NAME.equals(hostAddress) ? localAddressProvider.getLocalAddress() : hostAddress;
 
+    final SslContext sslContextClient;
+    final SslContext sslContextServer;
+
+    supportsSSL = System.getProperty("ssl") != null;
+
+    if (protocolType.equals(PROTOCOL_HTTP)) {
+      if (supportsSSL) {
+        try {
+          final SelfSignedCertificate ssc = new SelfSignedCertificate();
+          sslContextClient = SslContext.newServerContext(ssc.certificate(), ssc.privateKey());
+          sslContextServer = SslContext.newClientContext();
+          LOG.log(Level.FINE, "SSL context created");
+        } catch (final Exception ex) {
+          final RuntimeException transportException =
+              new TransportRuntimeException("Could not create SSL Context", ex);
+          LOG.log(Level.SEVERE, "Could not create SSL Context", ex);
+          throw transportException;
+        }
+      } else {
+        LOG.log(Level.FINE, "System property 'ssl' is not set");
+        sslContextClient = null;
+        sslContextServer = null;
+      }
+      this.uri = URI.create((supportsSSL ? "https://" : "http://") + hostAddress);
+    } else {
+      sslContextClient = null;
+      sslContextServer = null;
+      this.uri = null;
+    }
+
     this.numberOfTries = numberOfTries;
     this.retryTimeout = retryTimeout;
-    this.clientEventListener = new NettyClientEventListener(this.addrToLinkRefMap, clientStage);
-    this.serverEventListener = new NettyServerEventListener(this.addrToLinkRefMap, serverStage);
+    if (protocolType.equals(PROTOCOL_NETTY)) {
+      this.clientEventListener = new NettyClientEventListener(this.addrToLinkRefMap, clientStage);
```

Comment:
I've discussed it with @nhne in offline, and we decided not to refactor the constructor because there is a little benefit of refactoring it.